tests/test_class_sh_property.py

```diff
@@ -0,0 +1,163 @@
+# -*- coding: utf-8 -*-
+import pytest
+
+import env  # noqa: F401
+from pybind11_tests import class_sh_property as m
+
+
+@pytest.mark.xfail("env.PYPY", reason="gc after `del field` is apparently deferred")
+@pytest.mark.parametrize("m_attr", ("m_valu_readonly", "m_valu_readwrite"))
+def test_valu_getter(msg, m_attr):
+    # Reduced from PyCLIF test:
+    # https://github.com/google/clif/blob/c371a6d4b28d25d53a16e6d2a6d97305fb1be25a/clif/testing/python/nested_fields_test.py#L56
+    outer = m.Outer()
+    field = getattr(outer, m_attr)
+    assert field.num == -99
+    with pytest.raises(ValueError) as excinfo:
+        m.DisownOuter(outer)
+    assert msg(excinfo.value) == "Cannot disown use_count != 1 (loaded_as_unique_ptr)."
```

Comment:
The test case is clear and well defined behavior.

My concern is on the practical side -- it is hard to know where the reference to the object is. field could have been passed and kept in many remote part of the application.

A very canonical pattern I know of, for this type of situation is to mark the object as stale, and all uses of it errors. A very typical example is the 'NFS stale handle' error. Then the responsibility is distributed to the consumers of the inner objects, rather than the owner needing to reason a large code base 'who is reference my inner parts?'.